### DISCUSSION TITLE
Is RandomLinkSplit designed for Inductive or Transductive Link prediction?

### DISCUSSION THREAD

**Discussion by Xurorx** [pyg-team/pytorch_geometric#6360] [Q&A] [answered]
Hello,

I would want to implement an inductive link prediction task using RGCN on Heterogeneous graph.
I wonder if [RandomLinkSplit](https://pytorch-geometric.readthedocs.io/en/latest/_modules/torch_geometric/transforms/random_link_split.html#RandomLinkSplit)  is designed for inductive link prediction task? If not, what split function can I use?

Also, in the documentation of RandomLinkSplit, it is said "The split is performed such that the training split does not include edges in validation and test splits; and the validation split does not include edges in the test split.", however, I found this is not true. There is overlapping between positive training edges and positive validation edges, also there is overlapping between training message passing edges and positive training edges and etc.. Please see the below codes running on OGB-MAG dataset.

```
data = OGB_MAG(root='./data', preprocess='metapath2vec', transform=ToUndirected())[0]
data = data.to(device)
print(data)
```

HeteroData(
  paper={
    x=[736389, 128],
    year=[736389],
    y=[736389],
    train_mask=[736389],
    val_mask=[736389],
    test_mask=[736389]
  },
  author={ x=[1134649, 128] },
  institution={ x=[8740, 128] },
  field_of_study={ x=[59965, 128] },
  (author, affiliated_with, institution)={ edge_index=[2, 1043998] },
  (author, writes, paper)={ edge_index=[2, 7145660] },
  (paper, cites, paper)={ edge_index=[2, 10792672] },
  (paper, has_topic, field_of_study)={ edge_index=[2, 7505078] },
  (institution, rev_affiliated_with, author)={ edge_index=[2, 1043998] },
  (paper, rev_writes, author)={ edge_index=[2, 7145660] },
  (field_of_study, rev_has_topic, paper)={ edge_index=[2, 7505078] }
)

```
transform = RandomLinkSplit(
    num_val=0.1,
    num_test=0.1,
    edge_types=data.edge_types, 
    rev_edge_types=data.edge_types, 
    disjoint_train_ratio=0.3, 
    split_labels=True, 
    is_undirected=False
    )


train_data, val_data, test_data = transform(data)
```

```
train_edges = train_data.get_edge_store('author','affiliated_with', 'institution').edge_index
print(f"train_edges:\n {train_edges}")
train_edges_set = set(tuple(x) for x in train_edges.reshape(-1,2).numpy())

pos_train_edges = train_data.get_edge_store('author','affiliated_with', 'institution').pos_edge_label_index
print(f"pos_train_edges:\n {pos_train_edges}")
pos_train_edges_set = set(tuple(x) for x in pos_train_edges.reshape(-1,2).numpy())

print(f"length of train_edges set: {len(train_edges_set)}")
print(f"length of pos_train_edges set: {len(pos_train_edges_set)}")
print(f"difference between two sets: {len(train_edges_set.difference(pos_train_edges_set))}")
```

train_edges:
 tensor([[   3242,     305,    3227,  ...,     848,    2945,     651],
        [ 385848,  977934,  390426,  ...,   96514,  300828, 1118646]])
pos_train_edges:
 tensor([[453192, 690368,  49892,  ..., 656619, 658595,  90829],
        [  6976,   3105,   4668,  ...,   4467,   3549,   1126]])
length of train_edges set: 515947
length of pos_train_edges set: 231647
difference between two sets: 487971

```
val_edges = val_data.get_edge_store('author','affiliated_with', 'institution').edge_index
print(f"val_edges:\n {val_edges}")
val_edges_set = set(tuple(x) for x in val_edges.reshape(-1,2).numpy())

pos_val_edges = val_data.get_edge_store('author','affiliated_with', 'institution').pos_edge_label_index
print(f"pos_val_edges:\n {pos_val_edges}")
pos_val_edges_set = set(tuple(x) for x in pos_val_edges.reshape(-1,2).numpy())

print(f"length of val_edges set: {len(val_edges_set)}")
print(f"length of pos_val_edges set: {len(pos_val_edges_set)}")
print(f"difference between two sets: {len(val_edges_set.difference(pos_val_edges_set))}")
```

val_edges:
 tensor([[   6976,    3105,    4668,  ...,     848,    2945,     651],
        [ 453192,  690368,   49892,  ...,   96514,  300828, 1118646]])
pos_val_edges:
 tensor([[ 100574, 1007626,  887722,  ..., 1013388, 1088406, 1075442],
        [   1617,    1920,    2483,  ...,    4197,    6255,    4668]])
length of val_edges set: 719618
length of pos_val_edges set: 99769
difference between two sets: 702974

```
print(len(pos_train_edges_set))
print(len(pos_val_edges_set))
print(len(pos_train_edges_set.difference(pos_val_edges_set)))
```

231609
99819
222843

**Comment by rusty1s** [comment, marked as answer]
Are you sure you are creating the `set` correctly? Shouldn't it be
```python
train_edges_set = set(tuple(x) for x in train_edges.t().tolist())
```

**Reply by Xurorx** [reply]
Hi @rusty1s , thank you! 

I used your code, and it shows there is no overlapping between message passing edges and supervision edges. Also no overlapping between positive training edges and positive validation edges.

```
train_edges = train_data.get_edge_store('author','affiliated_with', 'institution').edge_index
print(f"train_edges:\n {train_edges}")
train_edges_set = set(tuple(x) for x in train_edges.t().tolist())

pos_train_edges = train_data.get_edge_store('author','affiliated_with', 'institution').pos_edge_label_index
print(f"pos_train_edges:\n {pos_train_edges}")
pos_train_edges_set = set(tuple(x) for x in pos_train_edges.t().tolist())

print(f"length of train_edges set: {len(train_edges_set)}")
print(f"length of pos_train_edges set: {len(pos_train_edges_set)}")
print(f"difference between two sets: {len(train_edges_set.difference(pos_train_edges_set))}")
```

train_edges:
 tensor([[   5823,    8398,     649,  ...,    8381,    7378,    6049],
        [ 534136,  881764,  988392,  ...,  371627, 1130002,  396120]])
pos_train_edges:
 tensor([[ 576568,  929760,  662618,  ..., 1094504,  734666,   20644],
        [    804,    2421,    6314,  ...,    2925,    1227,     651]])
length of train_edges set: 584640
length of pos_train_edges set: 250560
difference between two sets: 584640
